language/vm/src/file_format.rs

```diff
@@ -638,6 +666,36 @@ impl SignatureToken {
             ),
         }
     }
+
+    /// Creating a new type by Substituting the type variables with type actuals.
+    pub fn substitute(&self, tys: &[SignatureToken]) -> Result<SignatureToken, VMStaticViolation> {
+        use SignatureToken::*;
+
+        match self {
+            Bool => Ok(Bool),
+            U64 => Ok(U64),
+            String => Ok(String),
+            ByteArray => Ok(ByteArray),
+            Address => Ok(Address),
+            Struct(idx, actuals) => Ok(Struct(
+                *idx,
+                actuals
+                    .iter()
+                    .map(|ty| ty.substitute(tys))
+                    .collect::<Result<Vec<_>, _>>()?,
+            )),
+            Reference(ty) => Ok(Reference(Box::new(ty.substitute(tys)?))),
+            MutableReference(ty) => Ok(MutableReference(Box::new(ty.substitute(tys)?))),
+            TypeParameter(idx) => match tys.get(*idx as usize) {
+                Some(ty) => Ok(ty.clone()),
+                None => Err(VMStaticViolation::IndexOutOfBounds(
```

Comment:
As mentioned in the comment above, would it be possible to do this check only in the bounds checker?

Reply:
Yes, will do.